src/ReservoirRouter.sol

```diff
@@ -0,0 +1,291 @@
+        }
+        else {
+            uint amountBOptimal = ReservoirLibrary.quote(amountADesired, reserveA, reserveB);
+            if (amountBOptimal <= amountBDesired) {
+                require(amountBOptimal >= amountBMin, "RR: INSUFFICIENT_B_AMOUNT");
+                (amountA, amountB) = (amountADesired, amountBOptimal);
+            } else {
+                uint amountAOptimal = ReservoirLibrary.quote(amountBDesired, reserveB, reserveA);
+                assert(amountAOptimal <= amountADesired);
+                require(amountAOptimal >= amountAMin, "RR: INSUFFICIENT_A_AMOUNT");
+                (amountA, amountB) = (amountAOptimal, amountBDesired);
+            }
+        }
+    }
+
+    function addLiquidity(
+        address tokenA,
+        address tokenB,
+        uint256 curveId,
+        uint amountADesired,
+        uint amountBDesired,
+        uint amountAMin,
+        uint amountBMin,
+        address to
+    ) external payable returns (uint256 amountA, uint256 amountB, uint256 liquidity) {
+        address pair;
+        (amountA, amountB, pair) = _addLiquidity(tokenA, tokenB, curveId, amountADesired, amountBDesired, amountAMin, amountBMin);
+
+        _pay(tokenA, msg.sender, pair, amountA);
+        _pay(tokenB, msg.sender, pair, amountB);
+
+        liquidity = IReservoirPair(pair).mint(to);
+    }
+
+    function removeLiquidity(
+        address tokenA,
+        address tokenB,
+        uint256 curveId,
+        uint256 liquidity,
+        uint256 amountAMin,
+        uint256 amountBMin,
+        address to // when withdrawing to the router in UniV3, this address is 0. But we can just put the router's address
+    ) external /*payable*/ returns (uint256 amountA, uint256 amountB) {
+        address pair = ReservoirLibrary.pairFor(address(factory), tokenA, tokenB, curveId);
+        IReservoirPair(pair).transferFrom(msg.sender, pair, liquidity); // send liquidity to pair
+        (uint amount0, uint amount1) = IReservoirPair(pair).burn(to);
+
+        (address token0,) = ReservoirLibrary.sortTokens(tokenA, tokenB);
+        (amountA, amountB) = tokenA == token0 ? (amount0, amount1) : (amount1, amount0);
+
+        require(amountA >= amountAMin, "RR: INSUFFICIENT_A_AMOUNT");
+        require(amountB >= amountBMin, "RR: INSUFFICIENT_B_AMOUNT");
+    }
+
+    // **** SWAP ****
+    // requires the initial amount to have already been sent to the first pair
+    /// @param inOrOut true for exact in, false for exact out
+    function _swap(uint[] memory amounts, bool inOrOut, address[] memory path, uint256[] memory curveIds, address _to) internal {
+        for (uint i; i < path.length - 1; ) {
+            (address input, address output) = (path[i], path[i + 1]);
+            (address token0,) = ReservoirLibrary.sortTokens(input, output);
+            address to = i < path.length - 2 ? ReservoirLibrary.pairFor(address(factory), output, path[i + 2], curveIds[i + 1]) : _to;
```

Comment:
What's the gas cost of using pairFor in a loop?

Reply:
calling it 3 times in cost `218762`. Increases almost linearly with the number of calls

Reply:
Yeah, we will have to think of a solution to this, 60-70k gas is extreme

Reply:
I'll make an issue to track this, will resolve it in the near future

Reply:
created issue #2 to track this, will resolve it in the near future